File: Src/Syntax/Syntax/Exps.cs
using Citron.Infra;
using System;
using System.Collections.Generic;
using Citron.Collections;
using System.Linq;
using System.Security.Cryptography.X509Certificates;
using System.Text;

namespace Citron.Syntax;

public abstract record class Exp : ISyntaxNode
{
    public abstract TResult Accept<TVisitor, TResult>(ref TVisitor visitor)
        where TVisitor : struct, IExpVisitor<TResult>;
}

// 
public record class IdentifierExp(string Value, ImmutableArray<TypeExp> TypeArgs) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitIdentifier(this);
    
}

public record class StringExp(ImmutableArray<StringExpElement> Elements) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitString(this);
}

public record class IntLiteralExp(int Value) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitIntLiteral(this);
}

public record class BoolLiteralExp(bool Value) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitBoolLiteral(this);
}

// null
public record class NullLiteralExp : Exp
{
    public static readonly NullLiteralExp Instance = new NullLiteralExp();
    NullLiteralExp() { }

    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitNullLiteral(this);
}

public record class BinaryOpExp(BinaryOpKind Kind, Exp Operand0, Exp Operand1) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitBinaryOp(this);
}

public record class UnaryOpExp(UnaryOpKind Kind, Exp Operand) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitUnaryOp(this);
}

public record class CallExp(Exp Callable, ImmutableArray<Argument> Args) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitCall(this);
}

public partial record struct LambdaExpParam(TypeExp? Type, string Name, bool HasParams);

public record class LambdaExp(ImmutableArray<LambdaExpParam> Params, ImmutableArray<Stmt> Body) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitLambda(this);
}

// a[b]
public record class IndexerExp(Exp Object, Exp Index) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitIndexer(this);
}

public record class MemberExp(Exp Parent, string MemberName, ImmutableArray<TypeExp> MemberTypeArgs) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitMember(this);
}

public record class ListExp(ImmutableArray<Exp> Elems) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitList(this);
}

// new Type(2, 3, 4);
public record class NewExp(TypeExp Type, ImmutableArray<Argument> Args) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitNew(this);
}

// box i
public record class BoxExp(Exp InnerExp) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitBox(this);
}

// x is T
public record class IsExp(Exp Exp, TypeExp Type) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitIs(this);
}

// x as T
public record class AsExp(Exp Exp, TypeExp Type) : Exp
{
    public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitAs(this);
}

public interface IExpVisitor<TResult>
{
    TResult VisitIdentifier(IdentifierExp exp);
    TResult VisitString(StringExp exp);
    TResult VisitIntLiteral(IntLiteralExp exp);
    TResult VisitBoolLiteral(BoolLiteralExp exp);    
    TResult VisitNullLiteral(NullLiteralExp exp);
    TResult VisitBinaryOp(BinaryOpExp exp);
    TResult VisitUnaryOp(UnaryOpExp exp);
    TResult VisitCall(CallExp exp);
    TResult VisitLambda(LambdaExp exp);
    TResult VisitIndexer(IndexerExp exp);
    TResult VisitMember(MemberExp exp);
    TResult VisitList(ListExp exp);
    TResult VisitNew(NewExp exp);
    TResult VisitBox(BoxExp exp);
    TResult VisitIs(IsExp exp);
    TResult VisitAs(AsExp exp);
}

File: src/Syntax/Exps.cpp
#include "pch.h"
#include "Syntax/Exps.h"
#include "Syntax/Argument.h"
#include "Syntax/StringExpElements.h"
#include "Syntax/Stmts.h"

using namespace std;

namespace Citron::Syntax {

StringExp::StringExp(vector<StringExpElement> elements)
    : elements(std::move(elements)) { }

IMPLEMENT_DEFAULTS_DEFAULT(StringExp)

struct BinaryOpExp::Impl 
{
    BinaryOpKind kind;
    Exp operand0;
    Exp operand1;
};

BinaryOpExp::BinaryOpExp(BinaryOpKind kind, Exp operand0, Exp operand1)
    : impl(new Impl { kind, std::move(operand0), std::move(operand1) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(BinaryOpExp)

BinaryOpKind BinaryOpExp::GetKind()
{
    return impl->kind;
}

Exp& BinaryOpExp::GetOperand0()
{
    return impl->operand0;
}

Exp& BinaryOpExp::GetOperand1()
{
    return impl->operand1;
}

struct UnaryOpExp::Impl
{
    UnaryOpKind kind;
    Exp operand;
};

UnaryOpExp::UnaryOpExp(UnaryOpKind kind, Exp operand)
    : impl(new Impl { kind, std::move(operand) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(UnaryOpExp)

UnaryOpKind UnaryOpExp::GetKind()
{
    return impl->kind;
}

Exp& UnaryOpExp::GetOperand()
{
    return impl->operand;
}

struct CallExp::Impl
{
    Exp callable;
    vector<Argument> args;
};

CallExp::CallExp(Exp callable, vector<Argument> args)
    : impl(new Impl { std::move(callable), std::move(args)})
{
}

IMPLEMENT_DEFAULTS_PIMPL(CallExp)

Exp& CallExp::GetCallable()
{
    return impl->callable;
}

vector<Argument>& CallExp::GetArgs()
{
    return impl->args;
}

LambdaExp::LambdaExp(vector<LambdaExpParam> params, vector<Stmt> body)
    : params(std::move(params)), body(std::move(body)) { }

IMPLEMENT_DEFAULTS_DEFAULT(LambdaExp)

struct IndexerExp::Impl
{
    Exp obj;
    Exp index;
};

IndexerExp::IndexerExp(Exp obj, Exp index)
    : impl(new Impl { std::move(obj), std::move(index) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(IndexerExp)

Exp& IndexerExp::GetObject()
{
    return impl->obj;
}

Exp& IndexerExp::GetIndex()
{
    return impl->index;
}

struct MemberExp::Impl
{
    Exp parent;
    string memberName;
    vector<TypeExp> memberTypeArgs;
};

MemberExp::MemberExp(Exp parent, string memberName, vector<TypeExp> memberTypeArgs)
    : impl(new Impl { std::move(parent), std::move(memberName), std::move(memberTypeArgs) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(MemberExp)

Exp& MemberExp::GetParent()
{
    return impl->parent;
}

string& MemberExp::GetMemberName()
{
    return impl->memberName;
}

vector<TypeExp>& MemberExp::GetMemberTypeArgs()
{
    return impl->memberTypeArgs;
}

ListExp::ListExp(vector<Exp> elems)
    : elems(std::move(elems))
{

}

IMPLEMENT_DEFAULTS_DEFAULT(ListExp)

NewExp::NewExp(TypeExp type, vector<Argument> args)
    : type(std::move(type)), args(std::move(args)) { }

IMPLEMENT_DEFAULTS_DEFAULT(NewExp)


struct BoxExp::Impl
{
    Exp innerExp;
};

BoxExp::BoxExp(Exp innerExp)
    :impl(new Impl{ std::move(innerExp) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(BoxExp)

Exp& BoxExp::GetInnerExp()
{
    return impl->innerExp;
}

struct IsExp::Impl
{
    Exp exp;
    TypeExp type;
};

IsExp::IsExp(Exp exp, TypeExp type)
    : impl(new Impl { std::move(exp), std::move(type) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(IsExp)

Exp& IsExp::GetExp()
{
    return impl->exp;
}

TypeExp& IsExp::Type()
{
    return impl->type;
}

struct AsExp::Impl
{
    Exp exp;
    TypeExp type;
};

AsExp::AsExp(Exp exp, TypeExp type)
    : impl(new Impl { std::move(exp), std::move(type) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(AsExp)

Exp& AsExp::GetExp()
{
    return impl->exp;
}

TypeExp& AsExp::Type()
{
    return impl->type;
}

} // Citron::Syntax